File: lib/ex_scylla/session_builder.ex
defmodule ExScylla.SessionBuilder do
  alias ExScylla.Types, as: T
  alias ExScylla.Types.Errors.NewSessionError
  use ExScylla.Macros.Native, [
                          prefix: :sb,
                          docs_rs_path: "/scylla/transport/session_builder/struct.SessionBuilder.html"
                        ]

  native_f func: :default_execution_profile_handle,
         args: [sb, eph],
         args_spec: [T.session_builder(), T.execution_profile_handle()],
         return_spec: T.session_builder(),
         doc_example: """
         iex> sb = SessionBuilder.new()
         iex> sb = SessionBuilder.autoema_agreement_timeout(sb, 5_000)
         iex> true = is_reference(sb)
         """

  native_f func: :auto_schema_agreement_timeout,
           args: [sb, timeout_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.auto_schema_agreement_timeout(sb, 5_000)
           iex> true = is_reference(sb)
           """

  native_f_async func: :build,
                 args: [sb],
                 args_spec: [T.session_builder()],
                 return_spec: {:ok, T.session()} | {:error, NewSessionError.t()},
                 doc_example: """
                 iex> sb = SessionBuilder.new()
                 ...>      |> SessionBuilder.known_node("127.0.0.1:9042")
                 iex> {:ok, session} = SessionBuilder.build(sb)
                 iex> true = is_reference(session)
                 """

  native_f func: :compression,
           args: [sb, compression],
           args_spec: [T.session_builder(), T.transport_compression()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.compression(sb, :lz4)
           iex> sb = SessionBuilder.compression(sb, :snappy)
           iex> true = is_reference(sb)
           """

  native_f func: :connection_timeout,
           args: [sb, timeout_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.connection_timeout(sb, 5_000)
           iex> true = is_reference(sb)
           """

  native_f func: :disallow_shard_aware_port,
           args: [sb, disallow],
           args_spec: [T.session_builder(), boolean()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.disallow_shard_aware_port(sb, true)
           iex> true = is_reference(sb)
           """

  native_f func: :fetch_schema_metadata,
           args: [sb, fetch],
           args_spec: [T.session_builder(), boolean()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.fetch_schema_metadata(sb, true)
           iex> true = is_reference(sb)
           """


  native_f func: :keepalive_interval,
           args: [sb, interval_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.keepalive_interval(sb, 5_000)
           iex> true = is_reference(sb)
           """

  native_f func: :keepalive_timeout,
           args: [sb, timeout_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.keepalive_timeout(sb, 5_000)
           iex> true = is_reference(sb)
           """

  native_f func: :keyspaces_to_fetch,
           args: [sb, keyspaces],
           args_spec: [T.session_builder(), list(String.t())],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.keyspaces_to_fetch(sb, ["my_keyspace"])
           iex> true = is_reference(sb)
           """

  native_f func: :known_node,
           args: [sb, hostname],
           args_spec: [T.session_builder(), String.t()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.known_node(sb, "127.0.0.1:9042")
           iex> true = is_reference(sb)
           """

   native_f func: :known_node_addr,
            args: [sb, node_addr],
            args_spec: [T.session_builder(), {:inet.ip_address(), :inet.port_number()}],
            return_spec: T.session_builder(),
            doc_example: """
            iex> sb = SessionBuilder.new()
            iex> sb = SessionBuilder.known_node_addr(sb, {{127,0,0,1}, 9042})
            iex> true = is_reference(sb)
            """

  native_f func: :known_nodes,
           args: [sb, hostnames],
           args_spec: [T.session_builder(), [String.t()]],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.known_nodes(sb, ["127.0.0.1:9042", "127.0.0.2:9042"])
           iex> true = is_reference(sb)
           """

  native_f func: :known_nodes_addr,
           args: [sb, node_addrs],
           args_spec: [T.session_builder(), [{:inet.ip_address(), :inet.port_number()}]],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.known_nodes_addr(sb, [{{127,0,0,1}, 9042}, {{127,0,0,1}, 9042}])
           iex> true = is_reference(sb)
           """

  native_f func: :new,
           args: [],
           args_spec: [],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> true = is_reference(sb)
           """

  native_f func: :no_auto_schema_agreement,
           args: [sb],
           args_spec: [T.session_builder()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.no_auto_schema_agreement(sb)
           iex> true = is_reference(sb)
           """

  native_f func: :pool_size,
           args: [sb, size],
           args_spec: [T.session_builder(), T.pool_size()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.pool_size(sb, {:per_shard, 3})
           iex> sb = SessionBuilder.pool_size(sb, {:per_host, 10})
           iex> true = is_reference(sb)
           """

  native_f func: :refresh_metadata_on_auto_schema_agreement,
           args: [sb, refresh_metadata],
           args_spec: [T.session_builder(), boolean()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.refresh_metadata_on_auto_schema_agreement(sb, true)
           iex> true = is_reference(sb)
           """


  native_f func: :schema_agreement_interval,
           args: [sb, interval_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> interval_ms = 5_000
           iex> sb = SessionBuilder.schema_agreement_interval(sb, interval_ms)
           iex> true = is_reference(sb)
           """

  native_f func: :tcp_keepalive_interval,
           args: [sb, interval_ms],
           args_spec: [T.session_builder(), pos_integer()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> interval_ms = 5_000
           iex> sb = SessionBuilder.tcp_keepalive_interval(sb, interval_ms)
           iex> true = is_reference(sb)
           """

  native_f func: :tcp_nodelay,
           args: [sb, nodelay],
           args_spec: [T.session_builder(), boolean()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> sb = SessionBuilder.tcp_nodelay(sb, true)
           iex> true = is_reference(sb)
           """

  native_f func: :user,
           args: [sb, username, passwd],
           args_spec: [T.session_builder(), String.t(), String.t()],
           return_spec: T.session_builder(),
           doc_example: """
           iex> sb = SessionBuilder.new()
           iex> {username, passwd} = {"user", "myS3cr3tp@ssw0rd"}
           iex> sb = SessionBuilder.user(sb, username, passwd)
           iex> true = is_reference(sb)
           """
end

File: native/ex_scylla/src/session_builder.rs
pub mod types;
use rustler::env::{OwnedEnv, SavedTerm};
//use consistency::Consistency;
use super::runtime;
use crate::settings::pool_size::ScyllaPoolSize;
use rustler::types::atom;
use rustler::{Atom, Encoder, Env, NifResult, ResourceArc, Term};
use scylla::SessionBuilder;
use std::net::SocketAddr;
use std::num::NonZeroU32;
use std::sync::Mutex;
use std::time::Duration;
//use rustler::types::LocalPid;
use crate::types::*;
use crate::utils::*;
use std::cell::Cell;
use types::*;

macro_rules! use_builder {
    ($sbr:ident, $e:expr) => {
        let guard = $sbr.lock().unwrap();
        guard.set($e(guard.take()));
        drop(guard);
    };
}

// SesisonBuilder methods
#[rustler::nif]
fn sb_write_coalescing(
    sbr: ResourceArc<SessionBuilderResource>,
    enable: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.write_coalescing(enable) });
    sbr
}

#[rustler::nif]
fn sb_auto_schema_agreement_timeout(
    sbr: ResourceArc<SessionBuilderResource>,
    timeout_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.auto_schema_agreement_timeout(Duration::from_millis(timeout_ms))
    });
    sbr
}

#[rustler::nif]
fn sb_build<'a>(
    env: Env<'a>,
    opaque: Term<'a>,
    sbr: ResourceArc<SessionBuilderResource>,
) -> NifResult<Atom> {
    let sb: SessionBuilder = sbr.lock().unwrap().take();
    async_elixir!(env, opaque, {
        let res = sb.build().await;
        res.ex()
    });
    Ok(atom::ok())
}

#[rustler::nif]
fn sb_compression(
    sbr: ResourceArc<SessionBuilderResource>,
    compression: ScyllaTransportCompression,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.compression(Some(compression.into()))
    });
    sbr
}

#[rustler::nif]
fn sb_connection_timeout(
    sbr: ResourceArc<SessionBuilderResource>,
    timeout_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.connection_timeout(Duration::from_millis(timeout_ms))
    });
    sbr
}

/*
#[rustler::nif]
fn sb_default_execution_profile_handle(
    sbr: ResourceArc<SessionBuilderResource>,
    profile_handle: ExecutionProfileHandle,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.default_execution_profile_handle(profile_handle)
    });
    sbr
}
*/

#[rustler::nif]
fn sb_disallow_shard_aware_port(
    sbr: ResourceArc<SessionBuilderResource>,
    disallow: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.disallow_shard_aware_port(disallow)
    });
    sbr
}
#[rustler::nif]
fn sb_fetch_schema_metadata(
    sbr: ResourceArc<SessionBuilderResource>,
    fetch: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.fetch_schema_metadata(fetch)
    });
    sbr
}

/*
#[rustler::nif]
fn sb_host_filter(
    sbr: ResourceArc<SessionBuilderResource>,
    filter: Arc<dyn HostFilter>,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.host_filter(filter) });
    sbr
}
 */

#[rustler::nif]
fn sb_keepalive_interval(
    sbr: ResourceArc<SessionBuilderResource>,
    interval_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.keepalive_interval(Duration::from_millis(interval_ms))
    });
    sbr
}

#[rustler::nif]
fn sb_keepalive_timeout(
    sbr: ResourceArc<SessionBuilderResource>,
    timeout_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.keepalive_timeout(Duration::from_millis(timeout_ms))
    });
    sbr
}

#[rustler::nif]
fn sb_keyspaces_to_fetch(
    sbr: ResourceArc<SessionBuilderResource>,
    keyspaces: Vec<String>,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.keyspaces_to_fetch(keyspaces)
    });
    sbr
}

#[rustler::nif]
fn sb_known_node(
    sbr: ResourceArc<SessionBuilderResource>,
    hostname: String,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.known_node(hostname) });
    sbr
}

#[rustler::nif]
fn sb_known_node_addr(
    sbr: ResourceArc<SessionBuilderResource>,
    node_addr: ScyllaSocketAddr,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.known_node_addr(node_addr.into())
    });
    sbr
}

#[rustler::nif]
fn sb_known_nodes(
    sbr: ResourceArc<SessionBuilderResource>,
    hostnames: Vec<String>,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.known_nodes(hostnames.as_slice())
    });
    sbr
}

#[rustler::nif]
fn sb_known_nodes_addr(
    sbr: ResourceArc<SessionBuilderResource>,
    node_addrs: Vec<ScyllaSocketAddr>,
) -> ResourceArc<SessionBuilderResource> {
    let node_addrs: Vec<SocketAddr> = node_addrs.into_iter().map(|a| a.into()).collect();
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.known_nodes_addr(node_addrs.as_slice())
    });
    sbr
}

/*
#[rustler::nif]
fn sb_load_balancing(
    sbr: ResourceArc<SessionBuilderResource>,
    policy: ScyllaLoadBalancingPolicy,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.load_balancing(policy.r()) });
    sbr
}
*/

#[rustler::nif]
fn sb_new() -> ResourceArc<SessionBuilderResource> {
    ResourceArc::new(SessionBuilderResource(Mutex::new(Cell::new(
        SessionBuilder::new(),
    ))))
}
#[rustler::nif]
fn sb_no_auto_schema_agreement(
    sbr: ResourceArc<SessionBuilderResource>,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.no_auto_schema_agreement() });
    sbr
}
#[rustler::nif]
fn sb_pool_size(
    sbr: ResourceArc<SessionBuilderResource>,
    size: ScyllaPoolSize,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.pool_size(size.into()) });
    sbr
}

#[rustler::nif]
fn sb_refresh_metadata_on_auto_schema_agreement(
    sbr: ResourceArc<SessionBuilderResource>,
    refresh_metadata: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.refresh_metadata_on_auto_schema_agreement(refresh_metadata)
    });
    sbr
}

#[rustler::nif]
fn sb_schema_agreement_interval(
    sbr: ResourceArc<SessionBuilderResource>,
    interval_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.schema_agreement_interval(Duration::from_millis(interval_ms))
    });
    sbr
}

#[rustler::nif]
fn sb_tcp_keepalive_interval(
    sbr: ResourceArc<SessionBuilderResource>,
    interval_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.tcp_keepalive_interval(Duration::from_millis(interval_ms))
    });
    sbr
}

#[rustler::nif]
fn sb_tcp_nodelay(
    sbr: ResourceArc<SessionBuilderResource>,
    nodelay: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.tcp_nodelay(nodelay) });
    sbr
}

#[rustler::nif]
fn sb_tracing_info_fetch_attempts(
    sbr: ResourceArc<SessionBuilderResource>,
    attempts: u32,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.tracing_info_fetch_attempts(NonZeroU32::new(attempts).expect("attemts can not be 0"))
    });
    sbr
}

#[rustler::nif]
fn sb_tracing_info_fetch_consistency(
    sbr: ResourceArc<SessionBuilderResource>,
    consistency: ScyllaConsistency,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.tracing_info_fetch_consistency(consistency.into())
    });
    sbr
}

#[rustler::nif]
fn sb_tracing_info_fetch_interval(
    sbr: ResourceArc<SessionBuilderResource>,
    interval_ms: u64,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.tracing_info_fetch_interval(Duration::from_millis(interval_ms))
    });
    sbr
}
#[rustler::nif]
fn sb_use_keyspace(
    sbr: ResourceArc<SessionBuilderResource>,
    keyspace_name: String,
    case_sensitive: bool,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| {
        sb.use_keyspace(keyspace_name, case_sensitive)
    });
    sbr
}

#[rustler::nif]
fn sb_user(
    sbr: ResourceArc<SessionBuilderResource>,
    username: String,
    passwd: String,
) -> ResourceArc<SessionBuilderResource> {
    use_builder!(sbr, |sb: SessionBuilder| { sb.user(username, passwd) });
    sbr
}
